src/main/kotlin/io/hhplus/tdd/domain/point/PointUseCaseImpl.kt

```diff
@@ -0,0 +1,58 @@
+package io.hhplus.tdd.domain.point
+
+import io.hhplus.tdd.database.PointHistoryTable
+import io.hhplus.tdd.database.UserPointTable
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Component
+import java.util.concurrent.ConcurrentHashMap
+
+@Component
+class PointUseCaseImpl(
+    private val userPointTable: UserPointTable,
+    private val pointHistoryTable: PointHistoryTable,
+) : PointUseCase {
+    companion object {
+        private const val MAX_POINT = 1_000_000L
+    }
+
+    private val logger: Logger = LoggerFactory.getLogger(javaClass)
+
+    private val locks = ConcurrentHashMap<Long, Any>()
+
+    override fun getUserPoint(id: Long): UserPoint = userPointTable.selectById(id)
+
+    override fun charge(
+        id: Long,
+        amount: Long,
+    ): UserPoint {
+        if (amount <= 0) throw IllegalArgumentException("충전 금액은 0보다 커야합니다")
+        val lock = locks.computeIfAbsent(id) { Any() }
+        synchronized(lock) {
+            val current = userPointTable.selectById(id)
+            val newPoint = current.point + amount
```

Comment:
```suggestion
        val updated = current.charge(amount)
        userPointTable.insertOrUpdate(userId, updated.point)
```


```kotlin
        userPointTable.insertOrUpdate(userId, updated.point)
data class UserPoint(val id: Long, val point: Long, val updateMillis: Long) {
    fun charge(amount: Long): UserPoint {
        val total = this.point + amount
        require(total <= MAX_POINT) { "최대 잔고는 $MAX_POINT 입니다" }
        return this.copy(point = total, updateMillis = System.currentTimeMillis())
    }
}
```

이런식으로 포인트에 대한 계산 로직은 `UserPoint` 가 갖는건 어떨까요?